shared/actions/signup/index.js

```diff
@@ -156,219 +83,344 @@ function isValidName (name: ?string): ?string {
-    const {email, username, inviteCode, passphrase, deviceName} = getState().signup
-    paperKeyResponse = null
-
-    if (email && username && inviteCode && passphrase && deviceName) {
-      const params: signup_signup_rpc = {
-        method: 'signup.signup',
-        waitingHandler: isWaiting => dispatch(waiting(isWaiting)),
-        param: {
-          email,
-          inviteCode,
-          username,
-          deviceName,
-          passphrase: passphrase.stringValue(),
-          storeSecret: false,
-          skipMail
-        },
-        incomingCallMap: {
-          'keybase.1.loginUi.displayPrimaryPaperKey': ({sessionID, phrase}, response) => {
-            paperKeyResponse = response
-            dispatch({
-              type: Constants.showPaperKey,
-              payload: {paperkey: new HiddenString(phrase)}
-            })
-            onDisplayPaperKey && onDisplayPaperKey()
-            dispatch(nextPhase())
-          },
-          'keybase.1.gpgUi.wantToAddGPGKey': (params, {error, result}) => {
-            // Do not add a gpg key for now
-            result(false)
-          }
-        },
-        callback: (err, {passphraseOk, postOk, writeOk}) => {
-          if (err) {
-            console.error('error in signup:', err)
-            dispatch({
-              type: Constants.signup,
-              error: true,
-              payload: {signupError: new HiddenString(err + '')}
-            })
-            dispatch(nextPhase())
-            reject()
-          } else {
-            console.log('Successful signup', passphraseOk, postOk, writeOk)
-            resolve()
-          }
-        }
-      }
+function signupArgsSelector (state) {
+  const {email, username, inviteCode, passphrase, deviceName} = state.signup
+  return {email, username, inviteCode, passphrase, deviceName}
+}
 
-      engine.rpc(params)
-    } else {
-      console.error('Entered signup action with a null required field')
-      reject()
+// The real meat, this describes how the signup flow
+function * signupFlow (): any {
+  try {
+    // $FlowIssue
+    const {requestInvite} = yield race({
+      requestInvite: call(requestInviteSaga),
+      checkInviteCode: call(checkInviteCodeSaga)
+    })
+
+    // The user requested an invite we're done with the signup flow
+    if (requestInvite) {
+      return
     }
-  })
-}
 
-function waiting (isWaiting: boolean): SignupWaiting {
-  return {
-    type: Constants.signupWaiting,
-    payload: isWaiting
+    yield put(routeAppend('usernameAndEmail'))
+    yield call(checkUsernameEmailSaga)
+    yield put(routeAppend('passphraseSignup'))
+    yield call(checkPassphraseSaga)
+    yield put(routeAppend('deviceName'))
+    yield call(deviceNameSaga)
+    // $FlowIssue
+    const {email, username, inviteCode, passphrase, deviceName} = yield select(signupArgsSelector)
+    yield put(signupWithService(email, username, inviteCode, passphrase, deviceName, false))
+    yield fork(paperKeySaga)
```

Comment:
In retrospect we probably don't want a fork here.

It's probably best to avoid forks if possible